### PR TITLE
rewrite to use polling

### PR DESCRIPTION
**Pull request by jedwards1211**
This includes various API additions which I haven't documented yet.  Even though it's backwards-compatible, this should be a major release because the behavior differs in that it waits to see if the processes exited.

### CODE REVIEW

**Comment by jedwards1211** [comment]
Windows notes (Windows doesn't have real signals):
I don't think it's possible to tell which signal a process was killed with in windows
I don't think it's possible to intercept a signal and exit later in windows

**Comment by nelsonic** [comment]
@jedwards1211 if you have _time_ could you `fix` the merge conflict on this PR so we can _merge_ it in? thanks! ❤️

**Comment by jedwards1211** [comment]
@nelsonic done!

File: terminate.js [hunk 1]
```diff
@@ -3,20 +3,48 @@ var handlePsTreeCallback = require('./handlePsTreeCallback');
...
+ * exit before timing out and calling back with an error.
+ * @param {function=} callback (optional) - if you want to know once the
  * procesess have been terminated, supply a callback.
- *   @param {string} error - will be null if no error occured
- *   @param {boolean} done - will be true if the operation succeeded.
+ *   @param {Error} error - will be null if no error occured
  */
-
-module.exports = function terminate(pid, callback) {
+module.exports = function terminate(pid) {
```
Comment (nelsonic) [inline terminate.js:20]: @jedwards1211 there _appears_ to be only **one** parameter to the `terminate` function.
and yet there a whole bunch of _optional_ params in the Documentation. 

While I _personally_ (_really_) like the idea of using the [`arguments`](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Functions/arguments) Object and making all other arguments to the function _optional_, _some_ people prefer to keep all parameters _explicit_...

Also, for _clarity_, the _incrementing_ of the index should be done `after` the value/object/function has been extracted from `arguments`.

The _tests_ appear to work as _expected_: https://travis-ci.org/dwyl/terminate/jobs/190886264#L427
including the tests which pass in the `options` parameter. 

If someone `else` ever decides they have the time to ***refactor*** the code to use https://github.com/dwyl/goodparts as the "_style guide_" that would be a (_very_) welcome PR... see: https://github.com/dwyl/terminate/issues/29

File: terminate.js [hunk 2]
```diff
@@ -3,20 +3,48 @@ var handlePsTreeCallback = require('./handlePsTreeCallback');
...
   if(!pid) {
     throw new Error("No pid supplied to Terminate!")
   }
+
+  // extract optional arguments.  Any can be omitted, but the
+  // ones present have to be in the
+  var i = 1;
```
Comment (nelsonic) [inline terminate.js:27]: @jedwards1211 is there meant to be some sort of _loop_ here to check all the `arguments` to see if one of them is the `callback`?
e.g: if someone does `terminate(pid, callback, options)` what will happen...?

**Review by nelsonic — APPROVED** [review]
@jedwards1211 I _really_ ~~like~~ ***love*** what you have done here! 😍 
You've: 
+ [x] improved the module while maintaining backward compatibility for the API (_function call_).
+ [x] Adds tests which execute your new "branches".
+ [x] updates the usage example in the readme to reflect the simplified approach. 

The only thing I would _add_ is a description (_in the readme_) of the _optional_ parameters the function now accepts... see: https://github.com/dwyl/terminate/issues/30

I'm happy to _merge_ this because it adds value to Windows users https://github.com/dwyl/terminate/issues/26 while maintaining backward compatibility for _existing_ users... 

[](https://www.npmjs.com/package/terminate)

_Apparently_, a few people are using it... beats me! 🤔

**Comment by nelsonic** [comment]
@jedwards1211 `terminate@2.1.0` on NPM contains your latest changes! thanks again! ❤️  ✅  🚀

**Pull request merged**